core/src/sumeragi/network_topology.rs

```diff
@@ -128,47 +130,23 @@ impl Topology {
         &self.sorted_peers[self.min_votes_for_commit()]
     }
     /// Add or remove peers from the topology.
-    pub fn update_peer_list(&mut self, new_peer_list: &[PeerId]) {
-        let mut i = 0;
-        while i < self.sorted_peers.len() {
-            if new_peer_list.iter().any(|p| p == &self.sorted_peers[i]) {
-                i += 1;
-            } else {
-                self.sorted_peers.remove(i);
-            }
-        }
-        self.sorted_peers.extend(
-            new_peer_list
-                .iter()
-                .filter(|p| !self.sorted_peers.contains(p))
-                .cloned()
-                .collect::<Vec<PeerId>>(),
-        );
+    pub fn update_peer_list(&mut self, mut new_peers: HashSet<PeerId>) {
+        self.sorted_peers.retain(|peer| new_peers.remove(peer));
+        self.sorted_peers.extend(new_peers);
     }
     /// Rotate peers after each failed attempt to create a block.
     pub fn rotate_all(&mut self) {
         self.sorted_peers.rotate_left(1);
     }
     /// Re-arrange the set of peers after each successful block commit.
     pub fn rotate_set_a(&mut self) {
-        let top = self.sorted_peers.remove(0);
-        self.sorted_peers.insert(
-            self.min_votes_for_commit().min(self.sorted_peers.len()),
-            top,
-        );
+        let rotate_at = self.min_votes_for_commit().min(self.sorted_peers.len());
+        self.sorted_peers[..rotate_at].rotate_left(1);
     }
     /// Pull peers up in the topology to the top of the a set while preserving local order.
     pub fn lift_up_peers(&mut self, to_lift_up: &[PublicKey]) {
-        let mut observing = Vec::new();
-        let mut i = 0;
-        while i < self.sorted_peers.len() {
-            if to_lift_up.contains(&self.sorted_peers[i].public_key) {
-                i += 1;
-            } else {
-                observing.insert(0, self.sorted_peers.remove(i)); // This has to be insert(0) and not push in order to preserve order.
-            }
-        }
-        self.sorted_peers.extend(observing);
+        self.sorted_peers
+            .sort_by_cached_key(|peer| !to_lift_up.contains(&peer.public_key));
```

Comment:
previous leader always becomes the tail

Reply:
The test lift_up_peers demonstrates that the behavior is correct.

Reply:
you mean proxy tail? or last element in topology?

Reply:
sorry, I commented in the wrong place